File: converted_source/ubrstrat.pas.java
// unit Ubrstrat

from conversion_common import *;
import uclasses;
import delphi_compatability;

// const
kMutation = 0;
kWeight = 1;
kMaxBreedingSections = 30;
kFromFirstPlant = 0;
kFromSecondPlant = 1;
kFromProbabilityBasedOnWeightsForSection = 2;
kFromPlantWithGreaterWeightForSectionIfEqualFirstPlant = 3;
kFromPlantWithGreaterWeightForSectionIfEqualSecondPlant = 4;
kFromPlantWithGreaterWeightForSectionIfEqualChooseRandomly = 5;



class PdBreedingStrategy extends PdStreamableObject {
    public  mutationStrengths;
    public  firstPlantWeights;
    public short getNonNumericalParametersFrom;
    public boolean mutateAndBlendColorValues;
    public boolean chooseTdosRandomlyFromCurrentLibrary;
    
    public void create() {
        short i = 0;
        
        for (i = 0; i <= kMaxBreedingSections - 1; i++) {
            // defaults 
            // mutationStrengths default to zero 
            this.firstPlantWeights[i] = 50;
        }
        this.getNonNumericalParametersFrom = kFromPlantWithGreaterWeightForSectionIfEqualChooseRandomly;
        this.mutateAndBlendColorValues = true;
        this.chooseTdosRandomlyFromCurrentLibrary = true;
    }
    
    public void readFromFile(String aFileName) {
        TextFile inputFile = new TextFile();
        String inputLine = "";
        String lineType = "";
        KfStringStream stream = new KfStringStream();
        short i = 0;
        
        AssignFile(inputFile, aFileName);
        stream = null;
        try {
            Reset(inputFile);
            stream = UNRESOLVED.KfStringStream.create;
            UNRESOLVED.readln(inputFile, inputLine);
            if (UNRESOLVED.pos("STRATEGY", uppercase(inputLine)) <= 0) {
                throw new GeneralException.create("Improper format for breeding strategy file");
            }
            while (!UNRESOLVED.eof(inputFile)) {
                UNRESOLVED.readln(inputFile, inputLine);
                stream.onStringSeparator(inputLine, "=");
                lineType = stream.nextToken;
                stream.spaceSeparator;
                if (UNRESOLVED.pos("NUMERIC", uppercase(lineType)) > 0) {
                    this.getNonNumericalParametersFrom = StrToIntDef(stream.nextToken, kFromPlantWithGreaterWeightForSectionIfEqualChooseRandomly);
                } else if (UNRESOLVED.pos("MUTATION", uppercase(lineType)) > 0) {
                    for (i = 0; i <= kMaxBreedingSections - 1; i++) {
                        this.mutationStrengths[i] = StrToIntDef(stream.nextToken, 0);
                    }
                } else if (UNRESOLVED.pos("WEIGHT", uppercase(lineType)) > 0) {
                    for (i = 0; i <= kMaxBreedingSections - 1; i++) {
                        this.firstPlantWeights[i] = StrToIntDef(stream.nextToken, 50);
                    }
                } else if (UNRESOLVED.pos("COLORS", uppercase(lineType)) > 0) {
                    this.mutateAndBlendColorValues = UNRESOLVED.strToBool(stream.nextToken);
                } else if (UNRESOLVED.pos("3D OBJECTS", uppercase(lineType)) > 0) {
                    this.chooseTdosRandomlyFromCurrentLibrary = UNRESOLVED.strToBool(stream.nextToken);
                } else {
                    break;
                }
            }
        } finally {
            stream.free;
            CloseFile(inputFile);
        }
    }
    
    public void writeToFile(String aFileName) {
        TextFile outputFile = new TextFile();
        short i = 0;
        
        AssignFile(outputFile, aFileName);
        try {
            Rewrite(outputFile);
            writeln(outputFile, "[Breeding strategy]");
            write(outputFile, "Mutations=");
            for (i = 0; i <= kMaxBreedingSections - 1; i++) {
                write(outputFile, IntToStr(this.mutationStrengths[i]) + " ");
            }
            writeln(outputFile);
            write(outputFile, "Weights=");
            for (i = 0; i <= kMaxBreedingSections - 1; i++) {
                write(outputFile, IntToStr(this.firstPlantWeights[i]) + " ");
            }
            writeln(outputFile);
            writeln(outputFile, "Non-numeric=" + IntToStr(this.getNonNumericalParametersFrom));
            writeln(outputFile, "Vary colors=" + UNRESOLVED.boolToStr(this.mutateAndBlendColorValues));
            writeln(outputFile, "Vary 3D objects=" + UNRESOLVED.boolToStr(this.chooseTdosRandomlyFromCurrentLibrary));
        } finally {
            CloseFile(outputFile);
        }
    }
    
    // ------------------------------------------------------------------------- data transfer for binary copy 
    public void classAndVersionInformation(PdClassAndVersionInformationRecord cvir) {
        cvir.classNumber = uclasses.kPdBreedingStrategy;
        cvir.versionNumber = 0;
        cvir.additionNumber = 0;
    }
    
    public void streamDataWithFiler(PdFiler filer, PdClassAndVersionInformationRecord cvir) {
        short i = 0;
        
        super.streamDataWithFiler(filer, cvir);
        //FIX unresolved WITH expression: filer
        UNRESOLVED.streamSmallint(this.getNonNumericalParametersFrom);
        UNRESOLVED.streamBoolean(this.mutateAndBlendColorValues);
        UNRESOLVED.streamBoolean(this.chooseTdosRandomlyFromCurrentLibrary);
        for (i = 0; i <= kMaxBreedingSections - 1; i++) {
            UNRESOLVED.streamSmallint(this.mutationStrengths[i]);
        }
        for (i = 0; i <= kMaxBreedingSections - 1; i++) {
            UNRESOLVED.streamSmallint(this.firstPlantWeights[i]);
        }
    }
    
}

File: converted_source/ubrstrat.pas.py
# unit Ubrstrat

from conversion_common import *
import usupport
import uclasses
import usstream
import ufiler
import delphi_compatability

# const
kMutation = 0
kWeight = 1
kMaxBreedingSections = 30
kFromFirstPlant = 0
kFromSecondPlant = 1
kFromProbabilityBasedOnWeightsForSection = 2
kFromPlantWithGreaterWeightForSectionIfEqualFirstPlant = 3
kFromPlantWithGreaterWeightForSectionIfEqualSecondPlant = 4
kFromPlantWithGreaterWeightForSectionIfEqualChooseRandomly = 5

class PdBreedingStrategy(PdStreamableObject):
    def __init__(self):
        self.mutationStrengths = [0] * (range(0, kMaxBreedingSections + 1) + 1)
        self.firstPlantWeights = [0] * (range(0, kMaxBreedingSections + 1) + 1)
        self.getNonNumericalParametersFrom = 0
        self.mutateAndBlendColorValues = false
        self.chooseTdosRandomlyFromCurrentLibrary = false
    
    def create(self):
        i = 0
        
        for i in range(0, kMaxBreedingSections):
            # defaults 
            # mutationStrengths default to zero 
            self.firstPlantWeights[i] = 50
        self.getNonNumericalParametersFrom = kFromPlantWithGreaterWeightForSectionIfEqualChooseRandomly
        self.mutateAndBlendColorValues = true
        self.chooseTdosRandomlyFromCurrentLibrary = true
        return self
    
    def readFromFile(self, aFileName):
        inputFile = TextFile()
        inputLine = ""
        lineType = ""
        stream = KfStringStream()
        i = 0
        
        AssignFile(inputFile, aFileName)
        stream = None
        try:
            Reset(inputFile)
            stream = usstream.KfStringStream.create
            UNRESOLVED.readln(inputFile, inputLine)
            if UNRESOLVED.pos("STRATEGY", uppercase(inputLine)) <= 0:
                raise GeneralException.create("Improper format for breeding strategy file")
            while not UNRESOLVED.eof(inputFile):
                UNRESOLVED.readln(inputFile, inputLine)
                stream.onStringSeparator(inputLine, "=")
                lineType = stream.nextToken()
                stream.spaceSeparator()
                if UNRESOLVED.pos("NUMERIC", uppercase(lineType)) > 0:
                    self.getNonNumericalParametersFrom = StrToIntDef(stream.nextToken(), kFromPlantWithGreaterWeightForSectionIfEqualChooseRandomly)
                elif UNRESOLVED.pos("MUTATION", uppercase(lineType)) > 0:
                    for i in range(0, kMaxBreedingSections):
                        self.mutationStrengths[i] = StrToIntDef(stream.nextToken(), 0)
                elif UNRESOLVED.pos("WEIGHT", uppercase(lineType)) > 0:
                    for i in range(0, kMaxBreedingSections):
                        self.firstPlantWeights[i] = StrToIntDef(stream.nextToken(), 50)
                elif UNRESOLVED.pos("COLORS", uppercase(lineType)) > 0:
                    self.mutateAndBlendColorValues = usupport.strToBool(stream.nextToken())
                elif UNRESOLVED.pos("3D OBJECTS", uppercase(lineType)) > 0:
                    self.chooseTdosRandomlyFromCurrentLibrary = usupport.strToBool(stream.nextToken())
                else:
                    break
        finally:
            stream.free
            CloseFile(inputFile)
    
    def writeToFile(self, aFileName):
        outputFile = TextFile()
        i = 0
        
        AssignFile(outputFile, aFileName)
        try:
            Rewrite(outputFile)
            writeln(outputFile, "[Breeding strategy]")
            write(outputFile, "Mutations=")
            for i in range(0, kMaxBreedingSections):
                write(outputFile, IntToStr(self.mutationStrengths[i]) + " ")
            writeln(outputFile)
            write(outputFile, "Weights=")
            for i in range(0, kMaxBreedingSections):
                write(outputFile, IntToStr(self.firstPlantWeights[i]) + " ")
            writeln(outputFile)
            writeln(outputFile, "Non-numeric=" + IntToStr(self.getNonNumericalParametersFrom))
            writeln(outputFile, "Vary colors=" + usupport.boolToStr(self.mutateAndBlendColorValues))
            writeln(outputFile, "Vary 3D objects=" + usupport.boolToStr(self.chooseTdosRandomlyFromCurrentLibrary))
        finally:
            CloseFile(outputFile)
    
    # ------------------------------------------------------------------------- data transfer for binary copy 
    def classAndVersionInformation(self, cvir):
        cvir.classNumber = uclasses.kPdBreedingStrategy
        cvir.versionNumber = 0
        cvir.additionNumber = 0
    
    def streamDataWithFiler(self, filer, cvir):
        i = 0
        
        PdStreamableObject.streamDataWithFiler(self, filer, cvir)
        self.getNonNumericalParametersFrom = filer.streamSmallint(self.getNonNumericalParametersFrom)
        self.mutateAndBlendColorValues = filer.streamBoolean(self.mutateAndBlendColorValues)
        self.chooseTdosRandomlyFromCurrentLibrary = filer.streamBoolean(self.chooseTdosRandomlyFromCurrentLibrary)
        for i in range(0, kMaxBreedingSections):
            self.mutationStrengths[i] = filer.streamSmallint(self.mutationStrengths[i])
        for i in range(0, kMaxBreedingSections):
            self.firstPlantWeights[i] = filer.streamSmallint(self.firstPlantWeights[i])
    
